### PR TITLE
Initial implementation of multi app run for Kubernetes Dev 

### LINKED ISSUE #1324: [Proposal] Extend Multi app run for k8s in dev/test environment
## Describe the proposal
This is a proposal to extend multi app run `dapr run -f` to include k8s deployments in the dev/test environments. 

Goal: 
Run the given prebuilt user container with the right annotations and configurations where the initial values are picked from the multi-app run template which can also be used for self-hosted mode.  

## Release Note

RELEASE NOTE:

### PR DESCRIPTION
**Pull request by mukundansundar**
# Description

supersedes #1330 
closes #1324  

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #_[issue number]_

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation

### CODE REVIEW

File: cmd/init.go [hunk 1]
```diff
@@ -202,6 +204,7 @@ func init() {
 	defaultContainerRuntime := string(utils.DOCKER)
 
 	InitCmd.Flags().BoolVarP(&kubernetesMode, "kubernetes", "k", false, "Deploy Dapr to a Kubernetes cluster")
+	InitCmd.Flags().BoolVarP(&devMode, "dev", "", false, "Use Dev mode. Deploy Redis, Zipkin alson in the Kubernetes cluster")
```
Comment (pravinpushkar) [inline cmd/init.go:207]: nit: type in "alson"

File: pkg/kubernetes/logs.go [hunk 2]
```diff
@@ -14,17 +14,29 @@ limitations under the License.
...
+	v1 "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	"github.com/dapr/cli/pkg/print"
 )
 
 const (
 	daprdContainerName    = "daprd"
 	appIDContainerArgName = "--app-id"
+
+	maxListingRetry = 10
```
Comment (pravinpushkar) [inline pkg/kubernetes/logs.go:37]: Can we add some description here for these variables ?

**Review by pravinpushkar — COMMENTED** [review]
Suggested few changes.

File: pkg/kubernetes/kubernetes.go [hunk 3]
```diff
@@ -33,13 +33,26 @@ import (
...
+	dashboardReleaseName   = "dapr-dashboard"
+	thirdPartyDevNamespace = "default"
+	zipkinChartName        = "zipkin"
+	redisChartName         = "redis"
+	zipkinReleaseName      = "dapr-dev-zipkin"
+	redisReleaseName       = "dapr-dev-redis"
+	redisVersion           = "6.2"
+	bitnamiHelmRepo        = "https://charts.bitnami.com/bitnami"
+	daprHelmRepo           = "https://dapr.github.io/helm-charts"
+	zipkinHelmRepo         = "https://openzipkin.github.io/zipkin"
```
[the remark this answers is not in the document]
Comment (mukundansundar) [reply pkg/kubernetes/kubernetes.go:50]: not for this release. We can add it as an enhancement, based on feedback.

File: pkg/kubernetes/kubernetes.go [hunk 4]
```diff
@@ -60,7 +74,8 @@ type InitConfiguration struct {
 
 // Init deploys the Dapr operator using the supplied runtime version.
 func Init(config InitConfiguration) error {
-	err := installWithConsole(daprReleaseName, config.Version, "Dapr control plane", config)
+	helmRepoDapr := utils.GetEnv("DAPR_HELM_REPO_URL", daprHelmRepo)
```
[the remark this answers is not in the document]
Comment (mukundansundar) [reply pkg/kubernetes/kubernetes.go:77]: its used in the next line ... see again

File: pkg/standalone/run.go [hunk 5]
```diff
@@ -47,39 +47,46 @@ const (
 // RunConfig represents the application configuration parameters.
 type RunConfig struct {
```
Comment (mukundansundar) [inline pkg/standalone/run.go:48]: in a follow up PR, will move RunConfig out of Standalone package into its own package.

File: pkg/standalone/run.go [hunk 6]
```diff
@@ -213,6 +220,36 @@ func (config *RunConfig) Validate() error {
 	return nil
 }
 
+func (config *RunConfig) ValidateK8s() error {
```
Comment (mukundansundar) [inline pkg/standalone/run.go:223]: when RunConfig moves, this function will also move.

File: pkg/kubernetes/logs.go [hunk 7]
```diff
@@ -84,3 +99,99 @@ func Logs(appID, podName, namespace string) error {
...
+		}
+	}
+
+	return nil
+}
+
+func getPods(ctx context.Context, appID string, podClient v1.PodInterface) (*corev1.PodList, error) {
+	listCtx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	labelSelector := fmt.Sprintf("%s=%s", daprAppIDKey, appID)
+	fmt.Println("Select", labelSelector)
```
Comment (mukundansundar) [inline pkg/kubernetes/logs.go:188]: will remove in a follow up commit in this PR.

File: pkg/standalone/run.go (hunk 6 above)
Comment (pravinpushkar) [reply pkg/standalone/run.go:223]: I think we should put a description to this method or rename the method to somethings `validateRunConfigK8s` wdyt ?

File: pkg/runfileconfig/run_file_config.go [hunk 8]
```diff
@@ -59,6 +67,12 @@ func (a *App) GetLogsDir() string {
 	return logsPath
 }
 
+func (a *App) GetDeployDir() string {
```
Comment (pravinpushkar) [inline pkg/runfileconfig/run_file_config.go:70]: should put method description ?

**Comment by pravinpushkar** [comment]
LGTM.  Just a couple of nits.

**Review by pravinpushkar — APPROVED** [review]

**Comment by mukundansundar** [comment]
E2E for this in #1336

File: pkg/kubernetes/run.go [hunk 9]
```diff
@@ -13,24 +13,416 @@ limitations under the License.
...
+				},
+				Annotations: app.RunConfig.GetAnnotations(),
+			},
+			Spec: corev1.PodSpec{
+				Containers: []corev1.Container{
+					{
+						Name:            app.AppID,
+						Image:           app.ContainerImage,
+						Env:             getEnv(app),
+						ImagePullPolicy: corev1.PullAlways,
```
Comment (joebowbeer) [inline pkg/kubernetes/run.go:287]: @mukundansundar Can you this pull policy be configurable?

`Never` is more appropriate for locally built images.  (And `IfNotPresent` is the k8s default, except if latest.)

**Pull request merged**